lib/Transforms/Instrumentation/ExtendedPointer.cc

```diff
@@ -0,0 +1,409 @@
+  if (ErrorBB)
+    return ErrorBB;
+
+  Function *Fn = Inst->getParent()->getParent();
+  Module *Md = Fn->getParent();
+  LLVMContext &ctx = Md->getContext();
+  IRBuilder<>::InsertPointGuard Guard(*Builder);
+  ErrorBB = BasicBlock::Create(Fn->getContext(), "bound_overflow", Fn);
+  Builder->SetInsertPoint(ErrorBB);
+
+//  llvm::Value *F = Intrinsic::getDeclaration(Fn->getParent(), Intrinsic::trap);
+  AttrBuilder AB;
+  AB.addAttribute(Attribute::NoReturn);
+  AttributeList as = AttributeList::get(ctx, AttributeList::FunctionIndex, AB);
+  auto errorFn = dyn_cast<Function>(
+      Md->getOrInsertFunction("__VERIFIER_error", as, Type::getVoidTy(ctx)));
+  CallInst *TrapCall = Builder->CreateCall(errorFn);
+  TrapCall->setDoesNotReturn();
+  TrapCall->setDoesNotThrow();
+  TrapCall->setDebugLoc(Inst->getDebugLoc());
+  Builder->CreateUnreachable();
+
+  return ErrorBB;
+}
+
+
+/// emitBranchToTrap - emit a branch instruction to a trap block.
+/// If Cmp is non-null, perform a jump only if its value evaluates to true.
+void ExtendedPointer::emitBranchToTrap(Value *Cmp) {
+  // check if the comparison is always false
+  ConstantInt *C = dyn_cast_or_null<ConstantInt>(Cmp);
+  if (C) {
+    ++ChecksSkipped;
+    if (!C->getZExtValue())
+      return;
+    else
+      Cmp = nullptr; // unconditional branch
+  }
+  ++ChecksAdded;
+
+  BasicBlock::iterator Inst = Builder->GetInsertPoint();
+  BasicBlock *OldBB = Inst->getParent();
+  BasicBlock *Cont = OldBB->splitBasicBlock(Inst);
+  OldBB->getTerminator()->eraseFromParent();
+
+ if (Cmp)
+    BranchInst::Create(getErrorBB(), Cont, Cmp, OldBB);
+  else
+    BranchInst::Create(getErrorBB(), OldBB);
+
+}
+
+
+/// instrument - adds run-time bounds checks to memory accessing instructions.
+/// Ptr is the pointer that will be read/written, and InstVal is either the
+/// result from the load or the value being stored. It is used to determine the
+/// size of memory block that is touched.
+/// Returns true if any change was made to the IR, false otherwise.
+bool ExtendedPointer::instrument(Value *Ptr, Value *InstVal,
+                                const DataLayout &DL) {
+  uint64_t NeededSize = DL.getTypeStoreSize(InstVal->getType());
+  Value *NeededSizeVal = ConstantInt::get(IntptrTy, NeededSize);
+  DEBUG(dbgs() << "Instrument " << *Ptr << " for " << Twine(NeededSize)
+               << " bytes\n");
+
+  SizeOffsetEvalType SizeOffset = ObjSizeEval->compute(Ptr);
+  Value *Or;
+  if (!ObjSizeEval->bothKnown(SizeOffset)) {
+    WARN << "fatptr instrument " << *Ptr << " for " << Twine(NeededSize) << " bytes\n";
+    // get start and end by calling internalized functions
+    Value *Start = Builder->CreateCall(m_getFatSlot0, Builder->CreateBitCast(Ptr, Builder->getInt8PtrTy()));
+    Value *Size = Builder->CreateCall(m_getFatSlot1, Builder->CreateBitCast(Ptr, Builder->getInt8PtrTy()));
+    Value *recov = Builder->CreateCall(m_recoverFatPtr, Builder->CreateBitCast(Ptr, Builder->getInt8PtrTy()));
```

Comment:
needed for embedded pointer fat pointer impl (not needed for hashmap fat ptr impl)